src/Features/CSharp/Portable/CodeRefactorings/ConvertLocalFunctionToMethod/CSharpConvertLocalFunctionToMethodCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,230 @@
+        {
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var declaredSymbol = (IMethodSymbol)semanticModel.GetDeclaredSymbol(localFunction, cancellationToken);
+
+            var dataFlow = semanticModel.AnalyzeDataFlow(
+                localFunction.ExpressionBody?.Expression ?? (SyntaxNode)localFunction.Body);
+            var captures = dataFlow.Captured;
+            var capturesAsParameters = captures
+                .Where(x => !x.IsThisParameter())
+                .Select(capture => CodeGenerationSymbolFactory.CreateParameterSymbol(
+                    attributes: default,
+                    refKind: dataFlow.WrittenInside.Contains(capture) ? RefKind.Ref : default,
+                    isParams: default,
+                    type: capture.GetSymbolType(),
+                    name: capture.Name)).ToList();
+
+            var typeParameters = new List<ITypeParameterSymbol>();
+            GetCapturedTypeParameters(declaredSymbol, typeParameters);
+
+            var captureTypes = captures.Select(x => x.GetSymbolType()).OfType<ITypeParameterSymbol>().ToList();
+            RemoveUnusedTypeParameters(localFunction, semanticModel, captureTypes, typeParameters);
+
+            var container = localFunction.GetAncestor<MemberDeclarationSyntax>();
+            var containerSymbol = semanticModel.GetDeclaredSymbol(container, cancellationToken);
+            var isStatic = containerSymbol.IsStatic || captures.All(capture => !capture.IsThisParameter());
+
+            var methodName = GenerateUniqueMethodName(declaredSymbol);
+            var methodSymbol = CodeGenerationSymbolFactory.CreateMethodSymbol(
+                containingType: declaredSymbol.ContainingType,
+                attributes: default,
+                accessibility: Accessibility.Private,
+                modifiers: isStatic ? DeclarationModifiers.Static: default,
+                returnType: declaredSymbol.ReturnType,
+                refKind: default,
+                explicitInterfaceImplementations: default,
+                name: methodName,
+                typeParameters: typeParameters.ToImmutableArray(),
+                parameters: declaredSymbol.Parameters.AddRange(capturesAsParameters));
+
+            var method = MethodGenerator.GenerateMethodDeclaration(methodSymbol, CodeGenerationDestination.Unspecified,
+                    document.Project.Solution.Workspace, CodeGenerationOptions.Default, root.SyntaxTree.Options)
+                .WithExpressionBody(localFunction.ExpressionBody)
+                .WithSemicolonToken(localFunction.SemicolonToken)
+                .WithBody(localFunction.Body);
+
+            var generator = CSharpSyntaxGenerator.Instance;
+            var editor = new SyntaxEditor(root, generator);
+            editor.InsertAfter(container, method);
+            editor.RemoveNode(localFunction, SyntaxRemoveOptions.KeepNoTrivia);
```

Comment:
do you have a test where a local function recursively calls itself.  Can we have tests where you have this, and you also do:

1. renames.
2. type argument adding.
3. named arg adding.

Reply:
Done